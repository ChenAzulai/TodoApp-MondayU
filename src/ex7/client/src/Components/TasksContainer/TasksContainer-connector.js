import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { getFilteredItems } from "../../selectors/items-view-selectors";
import TasksContainer from "./TasksContainer";

const mapStateToProps = (state) => {
  const filteredTodoList = getFilteredItems(state);
  return { filteredTodoList };
};

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({}, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(TasksContainer);
