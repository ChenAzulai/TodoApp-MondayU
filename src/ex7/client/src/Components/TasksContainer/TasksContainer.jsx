import React from "react";
import TaskTicketConnector from "../TaskTicket/TaskTicket-connector";
import styles from "./TasksContainer.module.scss";
import PropTypes from "prop-types";

const TasksContainer = ({ todoList, searchInput, marked }) => {
  const hasSearch = searchInput.length !== 0;
  const hasMarkedFilter = marked.marked || marked.unmarked;

  const renderTasks = () => {
    let tasksToRender = todoList;
    if (hasSearch) {
      tasksToRender = todoList.filter((task) => {
        return searhProvider(task.itemName);
      });
    }
    if (hasMarkedFilter) {
      const isMarked = marked.marked;
      tasksToRender = tasksToRender.filter((task) => {
        return task.status === isMarked;
      });
    }
    return (
      <>
        {tasksToRender.map((item, index) => {
          return <TaskTicketConnector task={item} index={index} key={index} />;
        })}
      </>
    );
  };

  const searhProvider = (value) => {
    return value.toLowerCase().includes(searchInput.toLowerCase());
  };

  return (
    <div className={styles.task_container}>
      {renderTasks()}
    </div>
  );
};

TasksContainer.PropsTypes = {
  todoList: PropTypes.array,
  setIsLoading: PropTypes.func,
};

export default TasksContainer;
