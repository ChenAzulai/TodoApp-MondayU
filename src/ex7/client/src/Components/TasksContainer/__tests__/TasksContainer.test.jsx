import renderer from "react-test-renderer";
import TasksContainer from "../TasksContainer";
import { Provider } from "react-redux";
import { store } from "../../../store";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

const taskFactory = (tasks) => {
  if (tasks)
    return tasks.map((task) => {
      return { itemName: task, status: false };
    });
  else return [];
};

test("renders with no tasks", () => {
  const todoList = taskFactory();
  const marked = { marked: false, unmarked: false };
  const componentJSON = renderer
    .create(
      <Provider store={store}>
        <TasksContainer todoList={todoList} searchInput={""} marked={marked} />
      </Provider>
    )
    .toJSON();

  expect(componentJSON).toMatchSnapshot();
});

test("renders with 1 task", () => {
  const todoList = taskFactory(["task1"]);
  const marked = { marked: false, unmarked: false };
  const componentJSON = renderer
    .create(
      <Provider store={store}>
        <TasksContainer todoList={todoList} searchInput={""} marked={marked} />
      </Provider>
    )
    .toJSON();

  expect(componentJSON).toMatchSnapshot();
});

test("renders with multiple tasks", () => {
  const todoList = taskFactory(["task1", "task2", "task3", "task4"]);
  const marked = { marked: false, unmarked: false };
  const componentJSON = renderer
    .create(
      <Provider store={store}>
        <TasksContainer todoList={todoList} searchInput={""} marked={marked} />
      </Provider>
    )
    .toJSON();

  expect(componentJSON).toMatchSnapshot();
});

test("renders with no tasks and filterMarked is true", () => {
  const todoList = taskFactory();
  const marked = { marked: true, unmarked: false };
  const componentJSON = renderer
    .create(
      <Provider store={store}>
        <TasksContainer todoList={todoList} searchInput={""} marked={marked} />
      </Provider>
    )
    .toJSON();

  expect(componentJSON).toMatchSnapshot();
});

test("renders with no tasks and filterUnmarked is true", () => {
  const todoList = taskFactory();
  const marked = { marked: false, unmarked: true };
  const componentJSON = renderer
    .create(
      <Provider store={store}>
        <TasksContainer todoList={todoList} searchInput={""} marked={marked} />
      </Provider>
    )
    .toJSON();

  expect(componentJSON).toMatchSnapshot();
});

//Integration tests - send 2 items to the

const items = [
  {
    itemName: "Take dog out for a walk",
    status: false,
  },
  {
    itemName: "Do the dishes",
    status: true,
  },
];

//Integration tests - send 2 items to the
describe("TasksContainer with the items", () => {
  test("should render both items (one done and one not)", () => {
    render(
      <Provider store={store}>
        <TasksContainer filteredTodoList={items}/>
      </Provider>
    );
    const items1 = items[0];
    const items2 = items[1];
    const taskElement1 = screen.getByText(items1.itemName);
    const taskElement2 = screen.getByText(items2.itemName);

    expect(taskElement1).toBeInTheDocument();
    expect(taskElement2).toBeInTheDocument();
  });
});

