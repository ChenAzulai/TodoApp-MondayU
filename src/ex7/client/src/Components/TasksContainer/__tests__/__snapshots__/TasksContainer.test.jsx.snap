// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with 1 task 1`] = `
<div
  className="task_container"
/>
`;

exports[`renders with multiple tasks 1`] = `
<div
  className="task_container"
/>
`;

exports[`renders with no tasks 1`] = `
<div
  className="task_container"
/>
`;

exports[`renders with no tasks and filterMarked is true 1`] = `
<div
  className="task_container"
/>
`;

exports[`renders with no tasks and filterUnmarked is true 1`] = `
<div
  className="task_container"
/>
`;
